### PR TITLE
Create `with()` method to chain objects

### PR DESCRIPTION
**Pull request by slumbering**
Close to the issue: 
- https://github.com/dagger/dagger/issues/3885

This first iteration is using  the pattern proposed by @sipsma. 

```ts
func AddAFewMounts(c *dagger.Container) *dagger.Container {
  return c.
    WithMountedDirectory("/foo", c.Host().Directory("foo").
    WithMountedDirectory("/bar", c.Host().Directory("bar")
}

func main() {
 output, err := Container().
    From("alpine").
    WithWorkdir("/foo").
    With(AddAFewMounts).
    Exec("blah").Stdout().Contents()
...
}
```

@dolanor I know that you're doing some refactoring with the codegen so I'll leave you this part (talking about adding the `with()` method). You can directly push on this PR. Thanks

Signed-off-by: jffarge <jf@dagger.io>

### CODE REVIEW

File: sdk/nodejs/api/client.gen.ts
Comment (TomChv) [inline sdk/nodejs/api/client.gen.ts:684]: The arrow function is needed since we work on the `this,` but it creates inconsistency on the SDK.
Should we update every function to arrow style, or is this inconsistency allowable?

**Review by TomChv — COMMENTED** [review]
That's an excellent new feature! About code generation, I think each `type` that is composed of a query could have the method `with` that with the same content as in the PR.

**Why should we do this to all `types` composed of query?**

For consistency and to avoid the particular case, it's more complex to generate a `with` only for containers that generate a `with` for all types with queries in.

WDYT @dolanor @slumbering ?

File: sdk/nodejs/api/client.gen.ts
Comment (slumbering) [reply sdk/nodejs/api/client.gen.ts:684]: No you're right. It would be better to stay consistant. We should switch to arrow style on every method.

**Comment by slumbering** [comment]
> 



> That's an excellent new feature! About code generation, I think each `type` that is composed of a query could have the method `with` that with the same content as in the PR.
> 
> **Why should we do this to all `types` composed of query?**
> 
> For consistency and to avoid the particular case, it's more complex to generate a `with` only for containers that generate a `with` for all types with queries in.
> 
> WDYT @dolanor @slumbering ?

Agreed with that. Every types should have a `with()` method.

File: sdk/nodejs/api/client.gen.ts
Comment (TomChv) [reply sdk/nodejs/api/client.gen.ts:684]: Okay, I keep track of this in #4294

Comment (TomChv) [reply sdk/nodejs/api/client.gen.ts:684]: Oh, wait, I might have misunderstood the code.

The arrow function is used as a type, not as a value. The `with` function is still a classic function!

We're still consistent. My bad.

Comment (slumbering) [reply sdk/nodejs/api/client.gen.ts:684]: Yes the `with` function is still a classic function.

**Comment by slumbering** [comment]
@TomChv I'm aware that you've did a significant refacto [here](https://github.com/dagger/dagger/pull/4415/files) on the codegen, so forgive me for the `object.ts.tmpl` update ;)

**Comment by slumbering** [comment]
> That's an excellent new feature! About code generation, I think each `type` that is composed of a query could have the method `with` that with the same content as in the PR.
> 
> **Why should we do this to all `types` composed of query?**
> 
> For consistency and to avoid the particular case, it's more complex to generate a `with` only for containers that generate a `with` for all types with queries in.
> 
> WDYT @dolanor @slumbering ?

I've just updated the template to get the `with()` method for every types

**Comment by TomChv** [comment]
I updated this PR with an implementation of the Go `With` too!

@helderco do you think you can do it for python?

**Comment by helderco** [comment]
Instead of multiple people being assigned to the same PR, especially for different SDKs, I prefer separate issues/PRs.

I want to do it, but the when depends of backlog priorities.

**Comment by TomChv** [comment]
@helderco So can we merge this PR if it's supports correctly `With`?

**Comment by slumbering** [comment]
> Instead of multiple people being assigned to the same PR, especially for different SDKs, I prefer separate issues/PRs.
> 
> I want to do it, but the when depends of backlog priorities.

Agree with that. There is no rush and this is not a priority AFAIK. It was just a nice to have @TomChv.

**Comment by helderco** [comment]
> @helderco So can we merge this PR if it's supports correctly `With`?

I think so, at which point it can become a priority to keep the SDKs in sync, but it doesn't take long to implement.

**Comment by slumbering** [comment]
Concidering comments above and once checks passed, I'll merge this PR

**Comment by dolanor** [comment]
> **Why should we do this to all `types` composed of query?**
> 
> For consistency and to avoid the particular case, it's more complex to generate a `with` only for containers that generate a `with` for all types with queries in.
> 
> WDYT @dolanor @slumbering ?

I might not understand what are `types composed of query`, but I don't understand what a `secretA.with(secretB)` is supposed to give me? `secretA`? `secretB`? A merge of both? What would that look like?
For me, this pattern only applies to types who already use `withXxx()` methods, which for now are `Container` and `Directory`.
Am I missing something?

**Review by dolanor — CHANGES_REQUESTED** [review]
remove `with` for the types that don't need it. Unless I'm wrong.

**Comment by slumbering** [comment]
> pes composed of que



> > **Why should we do this to all `types` composed of query?**
> > For consistency and to avoid the particular case, it's more complex to generate a `with` only for containers that generate a `with` for all types with queries in.
> > WDYT @dolanor @slumbering ?
> 
> I might not understand what are `types composed of query`, but I don't understand what a `secretA.with(secretB)` is supposed to give me? `secretA`? `secretB`? A merge of both? What would that look like? For me, this pattern only applies to types who already use `withXxx()` methods, which for now are `Container` and `Directory`. Am I missing something?

The purpose of `.with()` is to group chain objects and is not related to methods (like Container and Directory) that have `withXxx()`. For instance you could use it with Git:

```ts
function gitFunc(c) {
  return c
  .branch('foo')
  .tag('bar')
  ...
}

connect(async (client) => {
  const src = client.git("http").with(gitFunc)
})
```

We've also choose [here](https://github.com/dagger/dagger/pull/4332#pullrequestreview-1241350348) to use `with()` in every methods for consistency.

**Comment by dolanor** [comment]
Ok, I think I see now. Good for me!

**Review by dolanor — APPROVED** [review]

**Pull request merged**
